File: src/Anemonis.VisualStudio.LayoutManager/Engine/LayoutPackagesProvider.cs
// © Alexander Kozlenko. Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

using Anemonis.VisualStudio.LayoutManager.Data;

using Newtonsoft.Json;

#pragma warning disable CA1822

namespace Anemonis.VisualStudio.LayoutManager.Engine
{
    /// <summary>Represents the layout packages provider.</summary>
    public sealed class LayoutPackagesProvider
    {
        private static readonly Regex _packageNameRegex = new Regex("^(?<i>[^,]+),version=(?<v>[^,]+)(?:,chip=(?<c>[^,]+))?(?:,language=(?<l>[^,]+))?$", RegexOptions.IgnoreCase | RegexOptions.Compiled);

        /// <summary>Initializes a new instance of the <see cref="LayoutPackagesProvider" /> class.</summary>
        public LayoutPackagesProvider()
        {
        }

        /// <summary>Gets catalog packages from the installation layout.</summary>
        /// <param name="json">The layout manifest JSON content.</param>
        /// <returns>The packages collection.</returns>
        /// <exception cref="ArgumentNullException"><paramref name="layoutPath" /> is <see langword="null" />.</exception>
        public IReadOnlyCollection<LayoutPackage> GetCatalogPackages(string json)
        {
            if (json == null)
            {
                throw new ArgumentNullException(nameof(json));
            }

            var catalog = JsonConvert.DeserializeObject<JsonLayoutCatalog>(json);
            var packages = new HashSet<LayoutPackage>(catalog.Packages.Length);

            for (var i = 0; i < catalog.Packages.Length; i++)
            {
                var package = new LayoutPackage(
                    catalog.Packages[i].Id,
                    catalog.Packages[i].Version,
                    catalog.Packages[i].Architecture,
                    catalog.Packages[i].Language
                );

                packages.Add(package);
            }

            return packages;
        }

        /// <summary>Gets actual packages from the installation layout.</summary>
        /// <param name="directories">The layout directories.</param>
        /// <returns>The packages collection.</returns>
        /// <exception cref="ArgumentNullException"><paramref name="layoutPath" /> is <see langword="null" />.</exception>
        public IReadOnlyCollection<LayoutPackage> GetLocalPackages(IReadOnlyList<string> directories)
        {
            if (directories == null)
            {
                throw new ArgumentNullException(nameof(directories));
            }

            var packages = new HashSet<LayoutPackage>(directories.Count);

            for (var i = 0; i < directories.Count; i++)
            {
                var match = _packageNameRegex.Match(Path.GetFileName(directories[i]));

                if (match.Success)
                {
                    var package = new LayoutPackage(
                        match.Groups["i"].Value,
                        match.Groups["v"].Value,
                        match.Groups["c"].Success ? match.Groups["c"].Value : null,
                        match.Groups["l"].Success ? match.Groups["l"].Value : null
                    );

                    packages.Add(package);
                }
            }

            return packages;
        }

        /// <summary>Gets obsolete packages from the installation layout.</summary>
        /// <param name="catalogPackages">The catalog packages from the installation layout.</param>
        /// <param name="localPackages">The local packages from the installation layout.</param>
        /// <returns>The packages collection.</returns>
        /// <exception cref="ArgumentNullException"><paramref name="catalogPackages" /> or <paramref name="localPackages" /> is <see langword="null" />.</exception>
        public IReadOnlyCollection<LayoutPackage> GetObsoletePackages(IReadOnlyCollection<LayoutPackage> catalogPackages, IReadOnlyCollection<LayoutPackage> localPackages)
        {
            if (catalogPackages == null)
            {
                throw new ArgumentNullException(nameof(catalogPackages));
            }
            if (localPackages == null)
            {
                throw new ArgumentNullException(nameof(localPackages));
            }

            return localPackages.Except(catalogPackages)
                .OrderBy(p => p.Id)
                .ThenBy(p => p.Version)
                .ThenBy(p => p.Architecture)
                .ThenBy(p => p.Language)
                .ToHashSet();
        }
    }
}

File: src/Anemonis.VisualStudio.LayoutManager/Engine/LayoutPackagesProvider.fs
// © Alexander Kozlenko. Licensed under the MIT License.

namespace Anemonis.VisualStudio.LayoutManager.Engine

open System
open System.Collections.Generic
open System.IO
open System.Text.Json
open System.Text.RegularExpressions

open Anemonis.VisualStudio.LayoutManager.Data

[<Sealed>]
type public LayoutPackagesProvider() =
    class
        static member private _packageNameRegex : Regex =
            new Regex("^(?<i>[^,]+),version=(?<v>[^,]+)(?:,chip=(?<c>[^,]+))?(?:,language=(?<l>[^,]+))?$", RegexOptions.IgnoreCase ||| RegexOptions.Compiled)

        member public this.GetCatalogPackages(json : string) : IReadOnlyCollection<LayoutPackage> =
            if json = null then
                raise (ArgumentNullException("json"))

            let catalog = JsonSerializer.Deserialize<JsonLayoutCatalog>(json)

            catalog.Packages
            |> Seq.map (fun x -> new LayoutPackage(x.ID, x.Version, x.Architecture, x.Language))
            |> HashSet<LayoutPackage>
            :> IReadOnlyCollection<LayoutPackage>

        member public this.GetLocalPackages(directories : IReadOnlyList<string>) : IReadOnlyCollection<LayoutPackage> =
            if directories = null then
                raise (ArgumentNullException("directories"))

            let getValue (g : Group) : string = (if g.Success then g.Value else null)

            directories
            |> Seq.map Path.GetFileName
            |> Seq.map LayoutPackagesProvider._packageNameRegex.Match
            |> Seq.where (fun x -> x.Success)
            |> Seq.map (fun x ->
                new LayoutPackage(
                    x.Groups.["i"].Value,
                    x.Groups.["v"].Value,
                    getValue(x.Groups.["c"]),
                    getValue(x.Groups.["l"])))
            |> HashSet<LayoutPackage>
            :> IReadOnlyCollection<LayoutPackage>

        member public this.GetObsoletePackages(catalogPackages : IReadOnlyCollection<LayoutPackage>, localPackages : IReadOnlyCollection<LayoutPackage>) : IReadOnlyCollection<LayoutPackage> =
            if catalogPackages = null then
                raise (ArgumentNullException("catalogPackages"))
            if localPackages = null then
                raise (ArgumentNullException("localPackages"))

            localPackages
            |> Seq.except catalogPackages
            |> Seq.sortBy (fun x -> x.ID, x.Version, x.Architecture, x.Language)
            |> HashSet<LayoutPackage>
            :> IReadOnlyCollection<LayoutPackage>
    end
